toolkit/tools/internal/pkggraph/pkggraph.go

```diff
@@ -1207,11 +1237,42 @@ func (g *PkgGraph) fixCycle(cycle []*PkgNode) (err error) {
 	metaNode := g.AddMetaNode(trimmedCycle, dependencyNodes)
 
 	// Enable cycle detection between meta nodes within the same srpm file
-	metaNode.SrpmPath = srpmPath
+	metaNode.SrpmPath = trimmedCycle[0].SrpmPath
 
 	return
 }
 
+// fixPrebuiltSRPMsCycle attempts to fix a cycle if at least one node is a pre-built SRPM.
+// If a cycle can be fixed, edges representing the build dependencies of the pre-built SRPM will be removed.
+func (g *PkgGraph) fixPrebuiltSRPMsCycle(trimmedCycle []*PkgNode) (err error) {
+	logger.Log.Debug("Checking if cycle contains pre-built SRPMs.")
+
+	previousNode := trimmedCycle[len(trimmedCycle)-1]
+	for _, currentNode := range trimmedCycle {
+		if isPrebuilt, _ := IsSRPMPrebuilt(currentNode.SrpmPath, g, nil); isPrebuilt {
+			logger.Log.Debugf("Found pre-built SRPM '%s' in cycle. Removing edge from '%s'.", previousNode.FriendlyName(), currentNode.FriendlyName())
+			g.RemoveEdge(previousNode.ID(), currentNode.ID())
+
+			goalNodes := g.findNodesByType(TypeGoal)
+			for _, goalNode := range goalNodes {
+				if g.HasEdgeFromTo(goalNode.ID(), currentNode.ID()) {
+					logger.Log.Debugf("Found edge from the '%s' goal node to the pre-built SRPM - no need for a new one.", goalNode.FriendlyName())
+					return
+				}
+			}
+
+			logger.Log.Debugf("No goal node depends on the pre-built SRPM. Adding edge from the '%s' goal node to ensure potential forced rebuild.", goalNodes[0].FriendlyName())
+			err = g.AddEdge(goalNodes[0], currentNode)
```

Comment:
I have a concern here, there is no guarantee that any goal nodes will survive the optimizer (unless that is expected?). The optimizer adds a new "PackagesToBuild" goal and creates a sub-graph from that.